homeassistant/components/home_connect/__init__.py

```diff
@@ -54,6 +126,118 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
+    async def _async_service_program(call, method):
+        """Execute calls to services taking a program."""
+        program = call.data[ATTR_PROGRAM]
+        entity_id = call.data[ATTR_ENTITY_ID]
+        options = {
+            ATTR_KEY: call.data.get(ATTR_KEY),
+            ATTR_VALUE: call.data.get(ATTR_VALUE),
+            ATTR_UNIT: call.data.get(ATTR_UNIT),
+        }
+        appliance = _get_appliance_by_entity_id(hass, entity_id)
+        if appliance is not None:
+            await hass.async_add_executor_job(
+                getattr(appliance, method), program, options
+            )
+
+    async def _async_service_command(call, command):
+        """Execute calls to services executing a command."""
+        entity_id = call.data[ATTR_ENTITY_ID]
+        appliance = _get_appliance_by_entity_id(hass, entity_id)
+        if appliance is not None:
+            await hass.async_add_executor_job(appliance.execute_command, command)
+
+    async def _async_service_key_value(call, method):
+        """Execute calls to services taking a key and value."""
+        key = call.data[ATTR_KEY]
+        value = call.data[ATTR_VALUE]
+        unit = call.data.get(ATTR_UNIT)
+        entity_id = call.data[ATTR_ENTITY_ID]
+        appliance = _get_appliance_by_entity_id(hass, entity_id)
+        if appliance is not None:
+            if unit is not None:
+                await hass.async_add_executor_job(
+                    getattr(appliance, method),
+                    key,
+                    value,
+                    unit,
+                )
+            else:
+                await hass.async_add_executor_job(
+                    getattr(appliance, method),
+                    key,
+                    value,
+                )
+
+    async def async_service_option_active(call):
+        """Service for setting an option for an active program."""
+        await _async_service_key_value(call, "set_options_active_program")
+
+    async def async_service_option_selected(call):
+        """Service for setting an option for a selected program."""
+        await _async_service_key_value(call, "set_options_selected_program")
+
+    async def async_service_setting(call):
+        """Service for changing a setting."""
+        await _async_service_key_value(call, "set_setting")
+
+    async def async_service_pause_program(call):
+        """Service for pausing a program."""
+        await _async_service_command(call, BSH_PAUSE)
+
+    async def async_service_resume_program(call):
+        """Service for resuming a paused program."""
+        await _async_service_command(call, BSH_RESUME)
+
+    async def async_service_select_program(call):
+        """Service for selecting a program."""
+        await _async_service_program(call, "select_program")
+
+    async def async_service_start_program(call):
+        """Service for starting a program."""
+        await _async_service_program(call, "start_program")
+
+    hass.services.async_register(
```

Comment:
These are all entity services, and should therefore be registered as such.

See: <https://developers.home-assistant.io/docs/dev_101_services#entity-services>

Reply:
Hi! Thank you for your feedback!

If I understand correctly, Entity services are for a particular platform. These services are used to manage settings and programs of a device, so they aren't specifically linked to an entity of the device. On which platform should I register these services?

Reply:
Yet, they all target a specific entity.

Reply:
Which entity would that be? I think the services target the device rather than a specific entity...

Reply:
I agree, except all these services accept entity IDs, not device. Hence I expect entity services instead

Reply:
The entity IDs are used to retrieve the device. But I think it is better to get the device by accepting the Home Connect appliance ID. Or is there a better way?

Reply:
> The entity IDs are used to retrieve the device. 

In that case, ask for devices, not entities. If you want to target entities, change the service.

Reply:
Is there some sort of ID for devices or should I just use Home Connects haID?

Reply:
Yes, we have that, device_id

Reply:
Thank you! Is that ID visible anywhere in the frontend?

Reply:
Well, not visible. But when using service calls, they show in a selector, same with creating automations in the UI.

Reply:
Thank you!